firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/MutationSquash.java

```diff
@@ -0,0 +1,119 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore.model.mutation;
+
+import static com.google.firebase.firestore.util.Assert.fail;
+import static com.google.firebase.firestore.util.Assert.hardAssert;
+
+import com.google.common.collect.Sets;
+import com.google.firebase.firestore.model.FieldPath;
+import com.google.firebase.firestore.model.MutableDocument;
+import com.google.firebase.firestore.model.ObjectValue;
+import com.google.firestore.v1.Value;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * Utility class to compute a mutation from an original document, and a mutated version. There are
+ * likely multiple mutations happened in between, hence "squash".
+ */
+public class MutationSquash {
+  public enum Type {
+    Set,
+    Delete,
+    Patch,
+    // Patch but with precondition set to None.
+    PatchWithMerge,
+    None
+  }
+
+  /**
+   * Give an original document, its mutated version, and the squash target type, returns a squashed
+   * mutation, that would hold invariant: original + squashed_mutation = finalDoc.
+   */
+  @Nullable
+  public static Mutation toMutation(MutableDocument original, MutableDocument finalDoc, Type type) {
+    switch (type) {
+      case Set:
+        return toSetMutation(original, finalDoc);
+      case Delete:
+        return toDeleteMutation(original, finalDoc);
+      case Patch:
+        return toPatchMutation(original, finalDoc, false);
+      case PatchWithMerge:
+        return toPatchMutation(original, finalDoc, true);
+      case None:
+        return null;
+      default:
+        fail("Unrecognized squash mutation type");
+    }
+    return null;
+  }
+
+  private static SetMutation toSetMutation(MutableDocument original, MutableDocument finalDoc) {
+    hardAssert(
+        original.getKey().equals(finalDoc.getKey()), "Squashing documents with different keys.");
+    return new SetMutation(finalDoc.getKey(), finalDoc.getData().clone(), Precondition.NONE);
+  }
+
+  private static DeleteMutation toDeleteMutation(
+      MutableDocument original, MutableDocument finalDoc) {
+    hardAssert(
+        original.getKey().equals(finalDoc.getKey()), "Squashing documents with different keys.");
+    return new DeleteMutation(finalDoc.getKey(), Precondition.NONE);
+  }
+
+  private static PatchMutation toPatchMutation(
+      MutableDocument original, MutableDocument finalDoc, boolean merge) {
+    hardAssert(
+        original.getKey().equals(finalDoc.getKey()), "Squashing documents with different keys.");
+
+    Map<String, Value> originalFields = original.getData().getFieldsMap();
+    Map<String, Value> finalFields = finalDoc.getData().getFieldsMap();
+    HashSet<FieldPath> mask = new HashSet<>();
+    HashMap<String, Value> values = new HashMap<>();
+
+    // Field updates
+    Sets.SetView<String> intersection =
```

Comment:
While this works, I think we are doing some extra work here. We apply the patch mutation to get the final document state, and then we go back to reverse engineer the diff. Do you think we can simplify this we we add this functionality to the Mutation class itself? Something like:

```
class Mutation {
   void applyToLocalView();
   void applyToRemoteDocument();
   Mutation mergeIntoExistingMutation(Mutation target);
}

Reply:
I opted for simplicity by trying to avoid doing merge while applying mutations (otherwise we need to handle merge(set, patch), merge(patch, patch), etc..I started that way and found it hard to get to a correct implementation), and also felt `applyToLocalView()` then call `mergeIntoExistingMutation` a bit too verbose.

If the main concern is performance, I think we can pass a FieldMask along the application chain, and only pick those fields to build a final mutation.

Reply:
I think I have two concerns: 
- You are essentially duplicating the mutation class hierarchy here and providing separate implementations for each mutation type (ignoring verify). This hides the coupling between the two implementations, even though these are actually tightly coupled. 
- The patch implementation here should be able to re-use private logic in PatchMutations. Wouldn't a PatchMutation applied to an empty document result in a squashed diff?